src/main/scala/chiseltest/formal/backends/TransitionSystemSimulator.scala

```diff
@@ -0,0 +1,273 @@
+          Memory(data.updated(ii.toInt, value))
+      }
+    }
+    def read(index: BigInt): BigInt = {
+      assert(index >= 0 && index < depth, s"index ($index) needs to be non-negative smaller than the depth ($depth)!")
+      data(index.toInt)
+    }
+    def ==(other: ArrayValue): Boolean = {
+      assert(other.isInstanceOf[Memory])
+      other.asInstanceOf[Memory].data == this.data
+    }
+  }
+  private def randomBits(bits: Int): BigInt = BigInt(bits, scala.util.Random)
+  private def randomSeq(depth: Int, bits: Int): IndexedSeq[BigInt] = (0 to depth).map(_ => randomBits(bits))
+
+  private def writesToMemory(depth: Int, bits: Int, writes: Iterable[(Option[BigInt], BigInt)]): Memory =
+    writes.foldLeft(Memory(randomSeq(depth, bits))) { case (mem, (index, value)) => mem.write(index, value) }
+
+  private val evalCtx: SMTEvalCtx = new SMTEvalCtx {
+    override def getBVSymbol(name: String): BigInt = {
+      val value = bvNameToIndex.get(name) match {
+        case Some(index) => data(index)
+        case None        => varData(name)
+      }
+      assert(value != null, s"Trying to read uninitialized symbol $name!")
+      value
+    }
+
+    override def getArraySymbol(name: String): Memory = memories(arrayNameToIndex(name))
+
+    override def startVariableScope(name: String, value: BigInt): Unit = {
+      assert(!varData.contains(name))
+      varData(name) = value
+    }
+    override def endVariableScope(name: String): Unit = {
+      varData.remove(name)
+    }
+
+    override def constArray(indexWidth: Int, value: BigInt) = {
+      Memory(IndexedSeq.fill(arrayDepth(indexWidth))(value))
+    }
+  }
+
+  private val CheckWidths = true
+  private def eval(expr: BVExpr): BigInt = {
+    val value = SMTExprEval.eval(expr)(evalCtx)
+    if (CheckWidths) {
+      val mask = (BigInt(1) << expr.width) - 1
+      if ((value & mask) != value) {
+        throw new RuntimeException(s"Failed to evaluate $expr!\nvalue $value does not fit into ${expr.width} bits!")
+      }
+    }
+    value
+  }
+  private def evalArray(expr: ArrayExpr): Memory = SMTExprEval.evalArray(expr)(evalCtx).asInstanceOf[Memory]
+
+  private def arrayDepth(indexWidth: Int): Int = (BigInt(1) << indexWidth).toInt
+
+  private def init(
+    regInit: Map[Int, BigInt],
+    memInit: Map[Int, Seq[(Option[BigInt], BigInt)]],
+    withVcd: Boolean
+  ): Unit = {
+    // initialize vcd
+    vcdWriter =
+      if (!withVcd) None
+      else {
+        val vv = vcd.VCD(sys.name)
+        vv.addWire("Step", 64)
+        allBV.foreach(s => vv.addWire(s.name, s.width))
+        observedMemories.foreach { m =>
+          val depth = arrayDepth(m.indexWidth)
+          (0 to depth).foreach(a => vv.addWire(s"${m.name}.$a", m.dataWidth))
+        }
+        Some(vv)
+      }
+
+    // initialize registers and memories in order (this ensures that they are topologically sorted by their init dependencies)
```

Comment:
I can't see `topological` without wondering if this might make use of `firrtl`'s `DiGraph` stuff. Just a question

Reply:
the actual sorting is done somewhere else :)